Sources/NIOExtras/LengthFieldBasedFrameDecoder.swift

```diff
@@ -0,0 +1,170 @@
+    
+    /// Create `LengthFieldBasedFrameDecoder` with a given frame length.
+    ///
+    /// - parameters:
+    ///    - lengthFieldLength: The length of the field specifying the remaining length of the frame.
+    ///    - lengthFieldEndianness: The endianness of the field specifying the remaining length of the frame.
+    ///
+    public init(lengthFieldLength: ByteLength, lengthFieldEndianness: Endianness = .big) {
+
+        // The value contained in the length field must be able to be represented by an integer type on the platform.
+        // ie. .eight == 64bit which would not fit into the Int type on a 32bit platform.
+        precondition(lengthFieldLength.length <= Int.bitWidth/8)
+            
+        self.lengthFieldLength = lengthFieldLength
+        self.lengthFieldEndianness = lengthFieldEndianness
+    }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        
+        if case .waitingForHeader = self.readState {
+            try self.readNextLengthFieldToState(buffer: &buffer)
+        }
+        
+        guard case .waitingForFrame(let frameLength) = self.readState else {
+            return .needMoreData
+        }
+        
+        guard let frameBuffer = try self.readNextFrame(buffer: &buffer, frameLength: frameLength) else {
+            return .needMoreData
+        }
+        
+        ctx.fireChannelRead(self.wrapInboundOut(frameBuffer))
+
+        return .continue
+    }
+    
+    private func readNextLengthFieldToState(buffer: inout ByteBuffer) throws {
+        
+        guard let lengthFieldSlice = buffer.readSlice(length: self.lengthFieldLength.length) else {
+            return
+        }
+        
+        // Convert the length field to an integer specifying the length
+        guard let lengthFieldValue = self.frameLength(for: lengthFieldSlice) else {
+            return
+        }
+        
+        self.readState = .waitingForFrame(length: lengthFieldValue)
+    }
+    
+    private func readNextFrame(buffer: inout ByteBuffer, frameLength: Int) throws -> ByteBuffer? {
+        
+        guard let contentsFieldSlice = buffer.readSlice(length: frameLength) else {
+            return nil
+        }
+        
+        self.readState = .waitingForHeader
+        
+        return contentsFieldSlice
+    }
+    
+    public func handlerRemoved(ctx: ChannelHandlerContext) {
+        if let buffer = cumulationBuffer, buffer.readableBytes > 0 {
+            ctx.fireErrorCaught(NIOExtrasErrors.LeftOverBytesError(leftOverBytes: buffer))
+        }
+    }
+
+    ///
+    /// Decodes the specified region of the buffer into an unadjusted frame length. The default implementation is
+    /// capable of decoding the specified region into an unsigned 8/16/32/64 bit integer.
+    /// - parameters:
+    ///    - buffer: The buffer containing the integer frame length
+    ///
+    private func frameLength(for buffer: ByteBuffer) -> Int? {
```

Comment:
regarding my comment above: I think the fix is really easy:
- just rename this function from `frameLength` to `readFrameLength`
- change all `getInteger` in there to `readInteger`
- remove the whole `guard let lengthFieldSlice = buffer.readSlice(length: self.lengthFieldLength.length) else { [...] }` block above and I think then it'll work already :)

Reply:
Yea, good point. I think that 2 stage check is a product of a few small refactorings and should probably go anyway, even discounting the fact that you have found a bug in it. Will fix.

Reply:
sounds great, thank you!